graveler/value.go

```diff
@@ -17,34 +15,23 @@ import (
  * ignored.
  */
 
-func writeVarint(w io.Writer, i int) error {
+func varintBytes(i int) []byte {
 	e := make([]byte, binary.MaxVarintLen64)
 	l := binary.PutVarint(e, int64(i))
-	_, err := w.Write(e[:l])
-	return err
+	return e[:l]
 }
 
-func writeBytes(tag string, w io.Writer, b []byte) error {
-	if err := writeVarint(w, len(b)); err != nil {
-		return fmt.Errorf("write %s length: %w", tag, err)
-	}
-	if _, err := w.Write(b); err != nil {
-		return fmt.Errorf("write %s: %w", tag, err)
-	}
-	return nil
+func putBytes(buf *[]byte, b []byte) {
+	*buf = append(*buf, varintBytes(len(b))...)
+	*buf = append(*buf, b...)
 }
 
 // MarshalValue returns bytes that uniquely unmarshal into a Value equal to v.
 func MarshalValue(v *Value) ([]byte, error) {
-	var w = &bytes.Buffer{}
-
-	if err := writeBytes("identity", w, v.Identity); err != nil {
-		return nil, err
-	}
-	if err := writeBytes("data", w, v.Data); err != nil {
-		return nil, err
-	}
-	return w.Bytes(), nil
+	ret := make([]byte, 0, len(v.Identity)+len(v.Data)+2*binary.MaxVarintLen32)
+	putBytes(&ret, v.Identity)
+	putBytes(&ret, v.Data)
+	return ret, nil
 }
```

Comment:
Suggesting the following, less pointers
```
// MarshalValue returns bytes that uniquely unmarshal into a Value equal to v.
func MarshalValue(v *Value) ([]byte, error) {
	ret := make([]byte, len(v.Identity)+len(v.Data)+2*binary.MaxVarintLen32)
	l := marshalVarintData(ret, v.Identity)
	l += marshalVarintData(ret[l:], v.Data)
	return ret[:l], nil
}

func marshalVarintData(out []byte, data []byte) int {
	dataLen := len(data)
	pos := binary.PutVarint(out, int64(dataLen))
	copy(out[pos:], data)
	pos += dataLen
	return pos
}```

Reply:
I'd rather not.  The code with fewer pointers looks much nicer but is brittle: the caller has to allocate a buffer large enough to hold the result.  Most of the ugliness of the existing code comes from having to call `append()`.  I think it is safer, hence worth the ugliness.  (It is surprisingly difficult to catch an error in the length calculation during `ret := make([]byte, ...)`: e.g. if you add a third value and say
```go
	ret := make([]byte, len(v.Identity)+len(v.Data)+len(v.Another)+2*binary.MaxVarintLen32)
```
then it will actually _work_ for all tests that use less than at least 1 GiB of data -- but it is still wrong.  I prefer the uglier code here :shrug: .